### ISSUE TITLE
Extend annotation processor API to allow input/output file specification

### ISSUE THREAD

**Issue by ewirch** [gradle/gradle#6440]
Some annotation processors use configuration files to allow control over their behavior. The annotation processor should be able to tell Gradle which files it uses as input.

F.e. Lombok has a `lombok.config`  which can significantly change the output of Lombok. Currently modifying the configuration file does not trigger a rebuild. Gradle considers the working tree as up to date. I filed a issue with Lombok: [#1817](https://github.com/rzwitserloot/lombok/issues/1817). The maintainer explained that the Gradle API is not rich enough to specify this kind of input file.

**Comment by oehme** [comment]
Well, long before an annotation processor can read something, Gradle needs to know that this file is an input to your task. Otherwise Gradle will determine that the task is up-to-date and not even ask Lombok to run. You can declare this using `compileJava.inputs.file 'path/to/lombok.config`.

Now I'm not sure whether this will actually make the task rerun. If not, we should fix that.

**Comment by ewirch** [comment]
Adding the configuration file as input of the `compileJava` task triggers a recompile on configuration file change.

So how would a general solution to this problem look like? Should each library, providing annotation processors with additional input, also provide a Gradle plugin, which will do the input file configuration?

**Comment by oehme** [comment]
Well, we could probably add some metadata file for processors so they can say "This option should be interpreted as an input file". That would be more convenient than making them write a Gradle plugin.

**Comment by hungvietnguyen** [comment]
(Sorry to jump in.) I worked with @lptr earlier to make data binding not break JavaCompile caching in the Android Gradle plugin (http://issuetracker.google.com/69243050). We used CommandLineArgumentProvider to annotate data binding resources as inputs/outputs. 

So maybe other annotation processors can do the same: AP authors should provide CommandLineArgumentProvider implementation classes out of the box. Users would then construct CommandLineArgumentProvider instances and pass that to Gradle or the Android Gradle plugin.

On the AGP side, we are working on a doc to explain to users how to do that.
